dvc/data/stage.py

```diff
@@ -27,21 +27,22 @@
 
 def _upload_file(from_fs_path, fs, odb, upload_odb, callback=None):
     from dvc.fs._callback import FsspecCallback
+    from dvc.fs.utils import fileobj_size
     from dvc.utils import tmp_fname
     from dvc.utils.stream import HashedStreamReader
 
     fs_path = upload_odb.fs.path
     tmp_info = fs_path.join(upload_odb.fs_path, tmp_fname())
-    with fs.open(from_fs_path, mode="rb") as stream:
-        stream = HashedStreamReader(stream)
-        size = fs.size(from_fs_path)
+    with fs.open(from_fs_path, mode="rb", chunk_size=fs.CHUNK_SIZE) as fobj:
+        stream = HashedStreamReader(fobj)
+        size = fileobj_size(fs, fobj, from_fs_path)
         with FsspecCallback.as_tqdm_callback(
             callback,
             desc=fs_path.name(from_fs_path),
             bytes=True,
             total=size,
         ) as cb:
-            upload_odb.fs.upload(stream, tmp_info, size=size, callback=cb)
+            upload_odb.fs.put_file(stream, tmp_info, size=size, callback=cb)
```

Comment:
We probably don't need to use `fs.utils.upload` here as it seems to upload to `tmp_info` already. WDYT @efiop?